File: ai/src/services/allocation.service.ts
/**
 * F1 — Guide Allocation & Workload Balancing
 *
 * Suggests optimal guide assignments for unassigned student groups
 * using a weighted scoring system that considers:
 *   - Available capacity (40%)
 *   - Domain match (30%)
 *   - Batch familiarity (20%)
 *   - Past supervision history (10%)
 */
import pool from '../config/db';
import { getMaxGuideLoad } from '../engine/thresholds';
import type { AllocationResult, GuideSuggestion } from '../types/ai.types';

// ── Get suggestions for all unassigned groups in a batch ──
export const getSuggestedAllocations = async (batchId: number): Promise<AllocationResult[]> => {
    const loadConfig = getMaxGuideLoad();

    // 1. Get unassigned groups
    const groupsRes = await pool.query(`
        SELECT g.id, g.group_name, p.domain
        FROM groups g
        LEFT JOIN projects p ON p.group_id = g.id
        WHERE g.batch_id = $1 AND g.guide_id IS NULL
        ORDER BY g.group_name
    `, [batchId]);

    if (groupsRes.rows.length === 0) return [];

    // 2. Get all available guides with their current loads
    const guidesRes = await pool.query(`
        SELECT u.uid AS guide_id, pr.full_name AS guide_name,
               COALESCE(pr.bio, '') AS specialization,
               COUNT(g.id)::int AS current_load
        FROM users u
        LEFT JOIN profiles pr ON pr.u_id = u.uid
        LEFT JOIN groups g ON g.guide_id = u.uid
        WHERE u.role = 'guide' AND u.is_deleted = FALSE
        GROUP BY u.uid, pr.full_name, pr.bio
        HAVING COUNT(g.id) < $1
        ORDER BY COUNT(g.id) ASC
    `, [loadConfig.hard_cap]);

    const guides = guidesRes.rows;

    // 3. For each guide, get batch familiarity info
    const batchGuideRes = await pool.query(`
        SELECT DISTINCT g.guide_id
        FROM groups g WHERE g.batch_id = $1 AND g.guide_id IS NOT NULL
    `, [batchId]);
    const batchFamiliarGuides = new Set(batchGuideRes.rows.map((r: any) => r.guide_id));

    // 4. Score each guide for each unassigned group
    const results: AllocationResult[] = [];

    for (const group of groupsRes.rows) {
        const suggestions: GuideSuggestion[] = [];

        for (const guide of guides) {
            const reasons: string[] = [];
            let score = 0;

            // Capacity score (weight: 40)
            const capacityScore = ((loadConfig.hard_cap - guide.current_load) / loadConfig.hard_cap) * 40;
            score += capacityScore;
            if (guide.current_load < loadConfig.default) {
                reasons.push(`Has capacity (${guide.current_load}/${loadConfig.hard_cap} groups)`);
            }

            // Domain match (weight: 30)
            if (group.domain && guide.specialization) {
                const domainLower = group.domain.toLowerCase();
                const specLower = guide.specialization.toLowerCase();
                if (specLower.includes(domainLower) || domainLower.includes(specLower)) {
                    score += 30;
                    reasons.push(`Domain match: "${group.domain}"`);
                }
            }

            // Batch familiarity (weight: 20)
            if (batchFamiliarGuides.has(guide.guide_id)) {
                score += 20;
                reasons.push('Already supervises groups in this batch');
            }

            // History bonus (weight: 10) — has supervised projects in similar domains before
            if (group.domain) {
                const historyRes = await pool.query(`
                    SELECT COUNT(*)::int AS count
                    FROM projects p
                    JOIN groups g ON g.id = p.group_id
                    WHERE g.guide_id = $1 AND LOWER(p.domain) = LOWER($2) AND p.status = 'approved'
                `, [guide.guide_id, group.domain]);
                if (historyRes.rows[0]?.count > 0) {
                    score += 10;
                    reasons.push(`Previously supervised ${historyRes.rows[0].count} project(s) in "${group.domain}"`);
                }
            }

            suggestions.push({
                guide_id: guide.guide_id,
                guide_name: guide.guide_name,
                current_load: guide.current_load,
                max_load: loadConfig.hard_cap,
                score: Math.round(score * 100) / 100,
                reasons,
            });
        }

        // Sort by score descending
        suggestions.sort((a, b) => b.score - a.score);

        results.push({
            group_id: group.id,
            group_name: group.group_name,
            suggestions: suggestions.slice(0, 5), // Top 5
        });
    }

    return results;
};

// ── Auto-allocate: assign top-scored guide to each group ──
export const autoAllocateGuides = async (batchId: number) => {
    const allocations = await getSuggestedAllocations(batchId);
    const assigned: Array<{ group_id: number; group_name: string; guide_id: number; guide_name: string }> = [];

    // Track load changes as we allocate
    const loadIncrements: Record<number, number> = {};

    for (const alloc of allocations) {
        if (alloc.suggestions.length === 0) continue;

        // Find best guide considering load increments from this batch run
        const best = alloc.suggestions.find(s => {
            const effectiveLoad = s.current_load + (loadIncrements[s.guide_id] || 0);
            return effectiveLoad < s.max_load;
        });

        if (!best) continue;

        await pool.query(`UPDATE groups SET guide_id = $1 WHERE id = $2`, [best.guide_id, alloc.group_id]);
        loadIncrements[best.guide_id] = (loadIncrements[best.guide_id] || 0) + 1;

        assigned.push({
            group_id: alloc.group_id,
            group_name: alloc.group_name,
            guide_id: best.guide_id,
            guide_name: best.guide_name,
        });
    }

    // Audit log
    await pool.query(`
        INSERT INTO ai_audit_log (action, details) VALUES ('auto_allocate_guides', $1)
    `, [JSON.stringify({ batch_id: batchId, assignments: assigned })]);

    return { assigned, total: assigned.length };
};

// ── Workload distribution stats ───────────────────────────
export const getWorkloadDistribution = async () => {
    const res = await pool.query(`
        SELECT u.uid AS guide_id, pr.full_name AS guide_name,
               COUNT(g.id)::int AS current_load,
               $1::int AS max_load
        FROM users u
        LEFT JOIN profiles pr ON pr.u_id = u.uid
        LEFT JOIN groups g ON g.guide_id = u.uid
        WHERE u.role = 'guide' AND u.is_deleted = FALSE
        GROUP BY u.uid, pr.full_name
        ORDER BY current_load DESC
    `, [getMaxGuideLoad().hard_cap]);

    const loads = res.rows.map((r: any) => r.current_load);
    const avg = loads.length > 0 ? loads.reduce((a: number, b: number) => a + b, 0) / loads.length : 0;
    const variance = loads.length > 0 ? loads.reduce((sum: number, l: number) => sum + Math.pow(l - avg, 2), 0) / loads.length : 0;
    const stdDev = Math.sqrt(variance);

    return {
        guides: res.rows,
        stats: {
            total_guides: res.rows.length,
            average_load: Math.round(avg * 100) / 100,
            std_deviation: Math.round(stdDev * 100) / 100,
            fairness_index: Math.round((1 - stdDev / Math.max(avg, 1)) * 100) / 100, // 1.0 = perfectly fair
        },
    };
};

File: ai-py/services/allocation_service.py
"""
F1 — Guide Allocation & Workload Balancing

Weighted scoring: capacity(40) + domain_match(30) + batch_familiarity(20) + history(10)
"""
import json
import math
from config.database import query, execute
from config.settings import get_max_guide_load


def get_suggested_allocations(batch_id: int) -> list:
    load_cfg = get_max_guide_load()
    groups = query(
        """SELECT g.id, g.group_name, p.domain
           FROM groups g LEFT JOIN projects p ON p.group_id = g.id
           WHERE g.batch_id = %s AND g.guide_id IS NULL ORDER BY g.group_name""",
        (batch_id,),
    )
    if not groups:
        return []

    guides = query(
        """SELECT u.uid AS guide_id, pr.full_name AS guide_name,
                  COALESCE(pr.bio, '') AS specialization, COUNT(g.id)::int AS current_load
           FROM users u LEFT JOIN profiles pr ON pr.u_id = u.uid
           LEFT JOIN groups g ON g.guide_id = u.uid
           WHERE u.role = 'guide' AND u.is_deleted = FALSE
           GROUP BY u.uid, pr.full_name, pr.bio
           HAVING COUNT(g.id) < %s ORDER BY COUNT(g.id) ASC""",
        (load_cfg["hard_cap"],),
    )
    familiar = {
        r["guide_id"]
        for r in query(
            "SELECT DISTINCT guide_id FROM groups WHERE batch_id = %s AND guide_id IS NOT NULL",
            (batch_id,),
        )
    }

    results = []
    for grp in groups:
        suggestions = []
        for g in guides:
            reasons, score = [], 0.0
            cap = ((load_cfg["hard_cap"] - g["current_load"]) / load_cfg["hard_cap"]) * 40
            score += cap
            if g["current_load"] < load_cfg["default"]:
                reasons.append(f"Has capacity ({g['current_load']}/{load_cfg['hard_cap']} groups)")

            if grp.get("domain") and g["specialization"]:
                if grp["domain"].lower() in g["specialization"].lower() or g["specialization"].lower() in grp["domain"].lower():
                    score += 30
                    reasons.append(f'Domain match: "{grp["domain"]}"')

            if g["guide_id"] in familiar:
                score += 20
                reasons.append("Already supervises groups in this batch")

            if grp.get("domain"):
                hist = query(
                    """SELECT COUNT(*)::int AS count FROM projects p
                       JOIN groups gr ON gr.id = p.group_id
                       WHERE gr.guide_id = %s AND LOWER(p.domain) = LOWER(%s) AND p.status = 'approved'""",
                    (g["guide_id"], grp["domain"]),
                )
                if hist and hist[0]["count"] > 0:
                    score += 10
                    reasons.append(f'Previously supervised {hist[0]["count"]} project(s) in "{grp["domain"]}"')

            suggestions.append({
                "guide_id": g["guide_id"], "guide_name": g["guide_name"],
                "current_load": g["current_load"], "max_load": load_cfg["hard_cap"],
                "score": round(score, 2), "reasons": reasons,
            })
        suggestions.sort(key=lambda x: x["score"], reverse=True)
        results.append({"group_id": grp["id"], "group_name": grp["group_name"], "suggestions": suggestions[:5]})
    return results


def auto_allocate_guides(batch_id: int) -> dict:
    allocations = get_suggested_allocations(batch_id)
    assigned, increments = [], {}
    for alloc in allocations:
        if not alloc["suggestions"]:
            continue
        best = None
        for s in alloc["suggestions"]:
            effective = s["current_load"] + increments.get(s["guide_id"], 0)
            if effective < s["max_load"]:
                best = s
                break
        if not best:
            continue
        execute("UPDATE groups SET guide_id = %s WHERE id = %s", (best["guide_id"], alloc["group_id"]))
        increments[best["guide_id"]] = increments.get(best["guide_id"], 0) + 1
        assigned.append({
            "group_id": alloc["group_id"], "group_name": alloc["group_name"],
            "guide_id": best["guide_id"], "guide_name": best["guide_name"],
        })
    execute("INSERT INTO ai_audit_log (action, details) VALUES ('auto_allocate_guides', %s)",
            (json.dumps({"batch_id": batch_id, "assignments": assigned}),))
    return {"assigned": assigned, "total": len(assigned)}


def get_workload_distribution() -> dict:
    load_cfg = get_max_guide_load()
    rows = query(
        """SELECT u.uid AS guide_id, pr.full_name AS guide_name, COUNT(g.id)::int AS current_load, %s AS max_load
           FROM users u LEFT JOIN profiles pr ON pr.u_id = u.uid LEFT JOIN groups g ON g.guide_id = u.uid
           WHERE u.role = 'guide' AND u.is_deleted = FALSE
           GROUP BY u.uid, pr.full_name ORDER BY current_load DESC""",
        (load_cfg["hard_cap"],),
    )
    loads = [r["current_load"] for r in rows]
    avg = sum(loads) / len(loads) if loads else 0
    variance = sum((l - avg) ** 2 for l in loads) / len(loads) if loads else 0
    std = math.sqrt(variance)
    return {
        "guides": rows,
        "stats": {
            "total_guides": len(rows), "average_load": round(avg, 2),
            "std_deviation": round(std, 2),
            "fairness_index": round(1 - std / max(avg, 1), 2),
        },
    }
